File: packages/totem-sdk/packages/tx-builder/src/coin-selection.ts
import type { KeyValueStorage, CoinFetcher, SpendableCoin } from './adapters.js';
export type { SpendableCoin } from './adapters.js';

export class CoinSelectionError extends Error {
  constructor(
    message: string,
    public readonly code: 'FETCH_FAILED' | 'INSUFFICIENT_FUNDS' | 'SERVICE_UNAVAILABLE' | 'NETWORK_ERROR',
    public readonly details?: Record<string, any>
  ) {
    super(message);
    this.name = 'CoinSelectionError';
  }
}

export interface CoinSelectionResult {
  selectedCoins: SpendableCoin[];
  totalSelected: string;
  change: string;
  insufficientFunds: boolean;
  fromAddresses: string[];
}

export type SendMode = 'global' | 'focused';

export interface CoinSelectionOptions {
  mode: SendMode;
  targetAmount: string;
  tokenId?: string;
  focusedAddress?: string;
  excludedAddresses?: string[];
}

const MINIMA_DECIMALS = 44;
const SCALE = BigInt(10) ** BigInt(MINIMA_DECIMALS);

const DECIMAL_RE = /^-?(?:0|[1-9][0-9]*)(?:\.[0-9]{1,44})?$/;

export function parseDecimalToBigInt(value: string): bigint {
  const clean = (value || '0').trim();
  if (!DECIMAL_RE.test(clean)) {
    throw new CoinSelectionError(`Invalid decimal format: "${value}"`, 'INSUFFICIENT_FUNDS');
  }
  const isNegative = clean.startsWith('-');
  const abs = isNegative ? clean.slice(1) : clean;
  const parts = abs.split('.');
  const intPart = parts[0] || '0';
  let fracPart = parts[1] || '';

  if (fracPart.length > MINIMA_DECIMALS) {
    fracPart = fracPart.slice(0, MINIMA_DECIMALS);
  } else {
    fracPart = fracPart.padEnd(MINIMA_DECIMALS, '0');
  }

  const result = BigInt(intPart) * SCALE + BigInt(fracPart);
  return isNegative ? -result : result;
}

export function bigIntToDecimalString(value: bigint): string {
  const isNegative = value < 0n;
  const absValue = isNegative ? -value : value;
  
  const intPart = absValue / SCALE;
  const fracPart = absValue % SCALE;
  
  const fracStr = fracPart.toString().padStart(MINIMA_DECIMALS, '0').replace(/0+$/, '');
  
  const sign = isNegative ? '-' : '';
  if (fracStr.length === 0) {
    return sign + intPart.toString();
  }
  return sign + intPart.toString() + '.' + fracStr;
}

export function addDecimalStrings(a: string, b: string): string {
  return bigIntToDecimalString(parseDecimalToBigInt(a) + parseDecimalToBigInt(b));
}

export function subtractDecimalStrings(a: string, b: string): string {
  return bigIntToDecimalString(parseDecimalToBigInt(a) - parseDecimalToBigInt(b));
}

export function compareDecimal(a: string, b: string): number {
  const bigA = parseDecimalToBigInt(a);
  const bigB = parseDecimalToBigInt(b);
  if (bigA < bigB) return -1;
  if (bigA > bigB) return 1;
  return 0;
}

export function isPositive(value: string): boolean {
  return parseDecimalToBigInt(value) > 0n;
}

export class CoinSelectionService {
  private excludedAddresses: Set<string> = new Set();
  private storage: KeyValueStorage | null;
  private fetcher: CoinFetcher;
  
  constructor(fetcher: CoinFetcher, storage?: KeyValueStorage) {
    this.fetcher = fetcher;
    this.storage = storage || null;
  }
  
  async loadExcludedAddresses(): Promise<void> {
    if (!this.storage) return;
    const data = await this.storage.get<string[]>('excludedAddresses');
    if (data && Array.isArray(data)) {
      this.excludedAddresses = new Set(data);
    }
  }
  
  async saveExcludedAddresses(): Promise<void> {
    if (!this.storage) return;
    await this.storage.set('excludedAddresses', Array.from(this.excludedAddresses));
  }
  
  addExcludedAddress(address: string): void {
    this.excludedAddresses.add(address);
    this.saveExcludedAddresses();
  }
  
  removeExcludedAddress(address: string): void {
    this.excludedAddresses.delete(address);
    this.saveExcludedAddresses();
  }
  
  getExcludedAddresses(): string[] {
    return Array.from(this.excludedAddresses);
  }
  
  isAddressExcluded(address: string): boolean {
    return this.excludedAddresses.has(address);
  }
  
  async fetchSpendableCoins(addresses: string[], tokenId: string = '0x00'): Promise<SpendableCoin[]> {
    return this.fetcher.fetchCoins(addresses, tokenId);
  }
  
  orderCoinsByAmount(coins: SpendableCoin[]): SpendableCoin[] {
    return [...coins].sort((a, b) => {
      const cmp = compareDecimal(b.amount, a.amount);
      if (cmp !== 0) return cmp;
      return a.coinId.localeCompare(b.coinId);
    });
  }
  
  selectCoins(
    coins: SpendableCoin[],
    options: CoinSelectionOptions
  ): CoinSelectionResult {
    let availableCoins = [...coins];
    
    if (options.mode === 'focused' && options.focusedAddress) {
      availableCoins = availableCoins.filter(c => c.address === options.focusedAddress);
    }
    
    const exclusions = options.excludedAddresses || this.getExcludedAddresses();
    if (exclusions.length > 0 && options.mode === 'global') {
      availableCoins = availableCoins.filter(c => !exclusions.includes(c.address));
    }
    
    if (options.tokenId && options.tokenId !== '0x00') {
      availableCoins = availableCoins.filter(c => c.tokenid === options.tokenId);
    }
    
    const orderedCoins = this.orderCoinsByAmount(availableCoins);
    
    const targetBigInt = parseDecimalToBigInt(options.targetAmount);
    let accumulatedBigInt = 0n;
    const selectedCoins: SpendableCoin[] = [];
    const fromAddresses = new Set<string>();
    
    for (const coin of orderedCoins) {
      if (accumulatedBigInt >= targetBigInt) break;
      
      selectedCoins.push(coin);
      accumulatedBigInt += parseDecimalToBigInt(coin.amount);
      fromAddresses.add(coin.address);
    }
    
    const totalSelected = bigIntToDecimalString(accumulatedBigInt);
    const insufficientFunds = accumulatedBigInt < targetBigInt;
    const change = insufficientFunds 
      ? '0' 
      : bigIntToDecimalString(accumulatedBigInt - targetBigInt);
    
    return {
      selectedCoins,
      totalSelected,
      change,
      insufficientFunds,
      fromAddresses: Array.from(fromAddresses)
    };
  }
  
  async selectCoinsForSend(
    allAddresses: string[],
    options: CoinSelectionOptions
  ): Promise<CoinSelectionResult> {
    let addressesToQuery: string[];
    
    if (options.mode === 'focused' && options.focusedAddress) {
      addressesToQuery = [options.focusedAddress];
    } else {
      const exclusions = options.excludedAddresses || this.getExcludedAddresses();
      addressesToQuery = allAddresses.filter(addr => !exclusions.includes(addr));
    }
    
    if (addressesToQuery.length === 0) {
      return {
        selectedCoins: [],
        totalSelected: '0',
        change: '0',
        insufficientFunds: true,
        fromAddresses: []
      };
    }
    
    const coins = await this.fetchSpendableCoins(addressesToQuery, options.tokenId);
    return this.selectCoins(coins, options);
  }
  
  formatCoinInputs(coins: SpendableCoin[]): Array<{ coinId: string }> {
    return coins.map(c => ({ coinId: c.coinId }));
  }
}

File: packages/totem-sdk/packages/tx-builder/rust/src/coin_selection.rs
use crate::types::*;
use crate::decimal::*;
use num_bigint::BigInt;
use num_traits::Zero;

pub fn order_coins_by_amount(coins: &[SpendableCoin]) -> Vec<SpendableCoin> {
    let mut sorted = coins.to_vec();
    sorted.sort_by(|a, b| {
        let cmp = compare_decimal(&b.amount, &a.amount).unwrap_or(0);
        use std::cmp::Ordering;
        match cmp {
            -1 => Ordering::Less,
            0 => a.coin_id.cmp(&b.coin_id),
            1 => Ordering::Greater,
            _ => Ordering::Equal,
        }
    });
    sorted
}

pub fn select_coins(
    coins: &[SpendableCoin],
    options: &CoinSelectionOptions,
    excluded_addresses: &[String],
) -> CoinSelectionResult {
    let mut available_coins: Vec<SpendableCoin> = coins.to_vec();

    if options.mode == "focused" {
        if let Some(ref focused_addr) = options.focused_address {
            available_coins.retain(|c| c.address == *focused_addr);
        }
    }

    let exclusions: Vec<&str> = if let Some(ref opt_exclusions) = options.excluded_addresses {
        opt_exclusions.iter().map(|s| s.as_str()).collect()
    } else {
        excluded_addresses.iter().map(|s| s.as_str()).collect()
    };

    if !exclusions.is_empty() && options.mode == "global" {
        available_coins.retain(|c| !exclusions.contains(&c.address.as_str()));
    }

    if let Some(ref token_id) = options.token_id {
        if token_id != "0x00" {
            available_coins.retain(|c| c.tokenid == *token_id);
        }
    }

    let ordered_coins = order_coins_by_amount(&available_coins);

    let target_bigint = parse_decimal_to_bigint(&options.target_amount)
        .unwrap_or_else(|_| BigInt::zero());
    let mut accumulated = BigInt::zero();
    let mut selected_coins: Vec<SpendableCoin> = Vec::new();
    let mut from_addresses: Vec<String> = Vec::new();

    for coin in &ordered_coins {
        if accumulated >= target_bigint {
            break;
        }

        let coin_amount = parse_decimal_to_bigint(&coin.amount)
            .unwrap_or_else(|_| BigInt::zero());
        selected_coins.push(coin.clone());
        accumulated += coin_amount;

        if !from_addresses.contains(&coin.address) {
            from_addresses.push(coin.address.clone());
        }
    }

    let total_selected = bigint_to_decimal_string(&accumulated);
    let insufficient_funds = accumulated < target_bigint;
    let change = if insufficient_funds {
        "0".to_string()
    } else {
        bigint_to_decimal_string(&(accumulated - target_bigint))
    };

    CoinSelectionResult {
        selected_coins,
        total_selected,
        change,
        insufficient_funds,
        from_addresses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_coin(id: &str, address: &str, amount: &str) -> SpendableCoin {
        SpendableCoin {
            coin_id: id.to_string(),
            address: address.to_string(),
            amount: amount.to_string(),
            tokenid: "0x00".to_string(),
            created: 0,
        }
    }

    #[test]
    fn test_select_exact() {
        let coins = vec![
            make_coin("c1", "addr1", "10"),
            make_coin("c2", "addr1", "5"),
            make_coin("c3", "addr2", "3"),
        ];
        let options = CoinSelectionOptions {
            mode: "global".to_string(),
            target_amount: "15".to_string(),
            token_id: None,
            focused_address: None,
            excluded_addresses: None,
        };
        let result = select_coins(&coins, &options, &[]);
        assert!(!result.insufficient_funds);
        assert_eq!(result.selected_coins.len(), 2);
        assert_eq!(result.total_selected, "15");
        assert_eq!(result.change, "0");
    }

    #[test]
    fn test_insufficient_funds() {
        let coins = vec![make_coin("c1", "addr1", "5")];
        let options = CoinSelectionOptions {
            mode: "global".to_string(),
            target_amount: "10".to_string(),
            token_id: None,
            focused_address: None,
            excluded_addresses: None,
        };
        let result = select_coins(&coins, &options, &[]);
        assert!(result.insufficient_funds);
    }

    #[test]
    fn test_change() {
        let coins = vec![
            make_coin("c1", "addr1", "10"),
            make_coin("c2", "addr1", "5"),
        ];
        let options = CoinSelectionOptions {
            mode: "global".to_string(),
            target_amount: "12".to_string(),
            token_id: None,
            focused_address: None,
            excluded_addresses: None,
        };
        let result = select_coins(&coins, &options, &[]);
        assert!(!result.insufficient_funds);
        assert_eq!(result.total_selected, "15");
        assert_eq!(result.change, "3");
    }

    #[test]
    fn test_focused_mode() {
        let coins = vec![
            make_coin("c1", "addr1", "10"),
            make_coin("c2", "addr2", "20"),
        ];
        let options = CoinSelectionOptions {
            mode: "focused".to_string(),
            target_amount: "5".to_string(),
            token_id: None,
            focused_address: Some("addr1".to_string()),
            excluded_addresses: None,
        };
        let result = select_coins(&coins, &options, &[]);
        assert_eq!(result.selected_coins.len(), 1);
        assert_eq!(result.selected_coins[0].coin_id, "c1");
    }

    #[test]
    fn test_excluded_addresses() {
        let coins = vec![
            make_coin("c1", "addr1", "10"),
            make_coin("c2", "addr2", "20"),
        ];
        let options = CoinSelectionOptions {
            mode: "global".to_string(),
            target_amount: "5".to_string(),
            token_id: None,
            focused_address: None,
            excluded_addresses: None,
        };
        let result = select_coins(&coins, &options, &["addr1".to_string()]);
        assert_eq!(result.selected_coins.len(), 1);
        assert_eq!(result.selected_coins[0].coin_id, "c2");
    }
}
